src/solvers/dg_unstructured_quad/dg_2d.jl

```diff
@@ -0,0 +1,341 @@
+# It constructs the basic `cache` used throughout the simulation to compute
+# the RHS etc.
+function create_cache(mesh::UnstructuredQuadMesh, equations::AbstractEquations,
+                      dg::DG, RealT, uEltype=RealT)
+
+  poly_deg = nnodes(dg.basis) - 1
+  nvars = nvariables(equations)
+
+  if poly_deg > mesh.poly_deg
+    error("polynomial degree of DG must be less than or equal to mesh polynomial degree")
+  end
+
+  elements = init_elements(RealT, mesh, dg.basis.nodes, nvars, poly_deg)
+
+  interfaces = init_interfaces(RealT, mesh, nvars, poly_deg)
+
+  if isperiodic(mesh)
+    boundaries = UnstructuredBoundaryContainer2D{RealT, nvars, poly_deg}(0)
+  else
+    boundaries = init_boundaries(RealT, mesh, elements, nvars, poly_deg)
+  end
+
+  cache = (; elements, interfaces, boundaries)
+
+  return cache
+end
+
+
+# Note! The mesh is appended to the end of some functions here such that we dispatch on the correct
+#       version and use the ::UnstructuredQuadMesh variable type below to keep track on it
+function rhs!(du::AbstractArray{<:Any,4}, u, t,
+              mesh::UnstructuredQuadMesh, equations,
+              initial_condition, boundary_conditions, source_terms,
+              dg::DG, cache)
+  # Reset du
+  @timeit_debug timer() "reset ∂u/∂t" du .= zero(eltype(du))
+
+  # Calculate volume integral
+  @timeit_debug timer() "volume integral" calc_contravariant_volume_integral!(du, u, have_nonconservative_terms(equations), equations,
+                                                                              dg.volume_integral, dg, cache)
+
+  # Prolong solution to interfaces
+  @timeit_debug timer() "prolong2interfaces" prolong2interfaces_right_handed!(cache, u, equations, dg)
+
+  # Calculate interface fluxes
+  @timeit_debug timer() "interface flux" calc_interface_flux_unstructured_quad!(cache.elements.surface_flux_values,
+                                                              have_nonconservative_terms(equations), equations,
+                                                              dg, cache)
+
+  # Prolong solution to boundaries
+  @timeit_debug timer() "prolong2boundaries" prolong2boundaries_right_handed!(cache, u, equations, dg)
+
+  # Calculate boundary fluxes
+  #  TODO: remove initial condition as an input argument here, only needed for hacky BCs
+  @timeit_debug timer() "boundary flux" calc_boundary_flux!(cache, t, boundary_conditions, equations, dg, initial_condition, mesh)
+
+  # Calculate surface integrals
+  @timeit_debug timer() "surface integral" calc_surface_integral_right_handed!(du, equations, dg, cache)
+
+  # Apply Jacobian from mapping to reference element
+  #  Note! this routine is reused from dg_curved/dg_2d.jl
+  @timeit_debug timer() "Jacobian" apply_jacobian!(du, mesh, equations, dg, cache)
+
+  # Calculate source terms
+  @timeit_debug timer() "source terms" calc_sources!(du, u, t, source_terms, equations, dg, cache)
+
+  return nothing
```

Comment:
Note to all: Since different function names are used here, it indicates that our abstraction is not good enough. I think it's better to pass the `mesh` explicitly here as done for the structured curved mesh and dispatch on this additional argument. That wouldn't be a breaking change since we don't `export` or document these functions such that they are not part of the official API.

Reply:
I second this :+1:

Reply:
okay, I originally dispatched on the `mesh` as the final argument but then as I started putting in comments (especially about the interface/surface stuff) I thought it might be necessary to distinguish what surface integral is being computed because the numbering convention of the sides is different

Reply:
I think if the mesh is properly documented, dispatching on the mesh is sufficient.